pkg/registry/registry.go

```diff
@@ -0,0 +1,316 @@
+package registry
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	k0sv1beta1 "github.com/k0sproject/k0s/pkg/apis/k0s/v1beta1"
+	clusterv1beta1 "github.com/replicatedhq/embedded-cluster-kinds/apis/v1beta1"
+	ectypes "github.com/replicatedhq/embedded-cluster-kinds/types"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// SeaweedfsS3SecretReadyConditionType represents the condition type that indicates status of
+	// the Seaweedfs secret.
+	SeaweedfsS3SecretReadyConditionType = "SeaweedfsS3SecretReady"
+
+	// RegistryS3SecretReadyConditionType represents the condition type that indicates status of
+	// the Registry secret.
+	RegistryS3SecretReadyConditionType = "RegistryS3SecretReady"
+)
+
+func EnsureSecrets(ctx context.Context, in *clusterv1beta1.Installation, metadata *ectypes.ReleaseMetadata, cli client.Client) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	config, op, err := ensureSeaweedfsS3Secret(ctx, in, metadata, cli)
+	if err != nil {
+		in.Status.SetCondition(metav1.Condition{
+			Type:               SeaweedfsS3SecretReadyConditionType,
+			Status:             metav1.ConditionFalse,
+			Reason:             "SecretFailed",
+			Message:            err.Error(),
+			ObservedGeneration: in.Generation,
+		})
+		return fmt.Errorf("ensure seaweedfs s3 secret: %w", err)
+	} else if op != controllerutil.OperationResultNone {
+		log.Info("Seaweedfs s3 secret changed", "operation", op)
+	}
+	in.Status.SetCondition(metav1.Condition{
+		Type:               SeaweedfsS3SecretReadyConditionType,
+		Status:             metav1.ConditionTrue,
+		Reason:             "SecretReady",
+		ObservedGeneration: in.Generation,
+	})
+
+	op, err = ensureRegistryS3Secret(ctx, in, metadata, cli, config)
+	if err != nil {
+		in.Status.SetCondition(metav1.Condition{
+			Type:               RegistryS3SecretReadyConditionType,
+			Status:             metav1.ConditionFalse,
+			Reason:             "SecretFailed",
+			Message:            err.Error(),
+			ObservedGeneration: in.Generation,
+		})
+		return fmt.Errorf("ensure registry s3 secret: %w", err)
+	} else if op != controllerutil.OperationResultNone {
+		log.Info("Registry s3 secret changed", "operation", op)
+	}
+	in.Status.SetCondition(metav1.Condition{
+		Type:               RegistryS3SecretReadyConditionType,
+		Status:             metav1.ConditionTrue,
+		Reason:             "SecretReady",
+		ObservedGeneration: in.Generation,
+	})
+
+	return nil
+}
+
+func ensureSeaweedfsS3Secret(ctx context.Context, in *clusterv1beta1.Installation, metadata *ectypes.ReleaseMetadata, cli client.Client) (*seaweedfsConfig, controllerutil.OperationResult, error) {
```

Comment:
nit pick: would it make more sense to move the seaweedfs related functions to seaweedfs.go in this package?